src/basic/util.c

```diff
@@ -832,6 +832,62 @@ uint64_t physical_memory_scale(uint64_t v, uint64_t max) {
         return r;
 }
 
+uint64_t system_tasks_max(void) {
+
+#if SIZEOF_PID_T == 4
+#define TASKS_MAX ((uint64_t) (INT32_MAX-1))
+#elif SIZEOF_PID_T == 2
+#define TASKS_MAX ((uint64_t) (INT16_MAX-1))
+#else
+#error "Unknown pid_t size"
+#endif
+
+        _cleanup_free_ char *value = NULL, *root = NULL;
+        uint64_t a = TASKS_MAX, b = TASKS_MAX;
+        int r;
+
+        /* Determine the maximum number of tasks that may run on this system. We check three source to determine this
+         * limit:
+         *
+         * a) the maximum value for the pid_t type
+         * b) the cgroups pids_max attribute for the system
+         * c) the kernel's configure maximum PID value
+         *
+         * And then pick the smallest of the three */
+
+        r = read_one_line_file("/proc/sys/kernel/pid_max", &value);
+        if (r < 0)
+                goto try_cgroups;
+
+        (void) safe_atou64(value, &a);
+
+try_cgroups:
+        if (cg_get_root_path(&root) < 0)
+                goto done;
+
+        value = mfree(value);
+        if (cg_get_attribute("pids", root, "pids.max", &value) < 0)
+                goto done;
+
+        (void) safe_atou64(value, &b);
+
+done:
+        return MIN3(TASKS_MAX, a, b);
+}
+
+uint64_t system_tasks_max_scale(uint64_t v, uint64_t max) {
+        uint64_t t, m;
+
+        t = system_tasks_max();
+        assert(t > 0);
+
+        m = t * v;
+        if (m / t != v) /* overflow? */
+                return UINT64_MAX;
+
+        return m / max;
```

Comment:
Ooh, I get it now. This scales by v/max, i. e. these act as numerator and denominator. 

The overflow case looks weird, though. E. g. if you want e. g. 80/100 and the *80 overflows, you'd get a grossly too large number (2^64-1) instead of a number which is at the same scale as the original `t`, and the caller does not get an error either. I think it would be safer and more predictable/sensible to just return `t` in the overflow case.